### PR TITLE
feat:[CCM-12293]: Anomaly detection for K8s including Harness Services

### PR DESCRIPTION
**Pull request by shashwat-harness**
## Describe your changes

Impact - Now we would be able to detect and monitor anomaly on services which was not present on the platform before

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52325)

### CODE REVIEW

**Comment by shashwat-harness** [comment]
trigger codeformat

**Comment by shashwat-harness** [comment]
trigger codeformat

**Comment by shashwat-harness** [comment]
trigger checkstylecodeformat

**Comment by shashwat-harness** [comment]
trigger uts

File: 450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java [hunk 1]
```diff
@@ -7,20 +7,7 @@
...
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.AZURE_RESOURCE_GROUP;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.AZURE_SUBSCRIPTION_GUID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.CLUSTER_NAME_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.GCP_PRODUCT_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.GCP_PROJECT_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.GCP_SKU_DESCRIPTION_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.NAMESPACE_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.REGION_FIELD_ID;
-import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.WORKLOAD_NAME_FIELD_ID;
+import static io.harness.ccm.commons.constants.ViewFieldLowerCaseConstants.*;
```
Comment (trunapushpa) [inline 450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java:10]: Harness codeformat check will fail for star imports. Import each one separately.

File: 490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java [hunk 2]
```diff
@@ -15,11 +15,11 @@ public interface ViewFieldConstants {
   String CLUSTER_TYPE_FIELD_ID = "clusterType";
   String NAMESPACE_FIELD_ID = "namespace";
   String WORKLOAD_NAME_FIELD_ID = "workloadName";
+  String SERVICE_NAME_FIELD_ID = "serviceName";
```
Comment (trunapushpa) [inline 490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java:18]: why move it from line 22 to line 18?

File: ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java [hunk 3]
```diff
@@ -39,6 +41,8 @@ public List<Pair<Integer, Class<? extends NGMigration>>> getMigrations() {
         .add(Pair.of(2, CeRecommendationTableAddJiraColumn.class))
         .add(Pair.of(3, CeRecommendationTableAddGovernanceColumn.class))
         .add(Pair.of(4, CCMRecommendationTableAddServicenowColumn.class))
+        .add(Pair.of(5, CCMAnomalyTableAddServiceColumn.class))
```
Comment (trunapushpa) [inline ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java:44]: merge both the migrations into one

File: 490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java (hunk 2 above)
Comment (shashwat-harness) [reply 490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java:18]: Nothing just aligning all important field constants together

File: 450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java (hunk 1 above)
Comment (shashwat-harness) [reply 450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java]: Done

File: ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java (hunk 3 above)
Comment (shashwat-harness) [reply ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java:43]: Merged

**Comment by shashwat-harness** [comment]
trigger checkstylecodeformat

**Comment by shashwat-harness** [comment]
trigger ut0

**Review by trunapushpa — CHANGES_REQUESTED** [review]

File: batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/AnomalyDetectionTimeSeries.java [hunk 4]
```diff
@@ -23,6 +24,7 @@
 @EqualsAndHashCode(callSuper = true)
 @SuperBuilder
 @Slf4j
+@Singleton
```
[the remark this answers is not in the document]
Comment (shashwat-harness) [reply batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/AnomalyDetectionTimeSeries.java]: Removed

File: batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/helpers/TimeSeriesUtils.java [hunk 5]
```diff
@@ -20,7 +20,9 @@
 import java.util.List;
 import java.util.stream.Collectors;
 import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
+@Slf4j
```
[the remark this answers is not in the document]
Comment (shashwat-harness) [reply batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/helpers/TimeSeriesUtils.java:25]: Done

File: batch-processing/service/src/main/java/io/harness/batch/processing/tasklet/support/HarnessEntitiesService.java [hunk 6]
```diff
@@ -25,6 +25,10 @@ public enum HarnessEntities { APP, SERVICE, ENV }
   public static class CacheKey {
     public String entityId;
     public HarnessEntities entity;
+    public CacheKey(String entityId, HarnessEntities entity) {
```
[the remark this answers is not in the document]
Comment (shashwat-harness) [reply batch-processing/service/src/main/java/io/harness/batch/processing/tasklet/support/HarnessEntitiesService.java:28]: Yes, but i don't think it provides any added advantage or make a difference

**Comment by shashwat-harness** [comment]
trigger uts

**Review by trunapushpa — DISMISSED** [review]

**Comment by shashwat-harness** [comment]
trigger uts

**Comment by shashwat-harness** [comment]
trigger uts

**Comment by shashwat-harness** [comment]
trigger uts

**Comment by shashwat-harness** [comment]
trigger ut0

**Comment by shashwat-harness** [comment]
trigger ut1

**Comment by shashwat-harness** [comment]
trigger ut5

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut8

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut0

**Comment by shashwat-harness** [comment]
trigger ut1

**Comment by shashwat-harness** [comment]
trigger ut1

**Comment by shashwat-harness** [comment]
trigger ut9

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut8

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut5

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut1

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut6

**Comment by shashwat-harness** [comment]
trigger ut0

**Comment by shashwat-harness** [comment]
trigger ut1

**Comment by shashwat-harness** [comment]
trigger ut1

**Review by trunapushpa — APPROVED** [review]

**Review by trunapushpa — APPROVED** [review]
LGTM

> Reviewed 25 of 40 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52325#-:-NdkIT2C1VlTQX_Xt8iK:bb6ywmx)* status: :shipit: complete! ✓

___
*[`450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java` line 10 at r1](https://reviewable.io/reviews/harness/harness-core/52325#gh-1312277257:-NdkHz30-dNEx308o-lr:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/ba7e82d2d5e1c11742dfd3fe02ef214afea4b867/450-ce-views/src/main/java/io/harness/ccm/views/helper/PerspectiveToAnomalyQueryHelper.java#L10)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Done
</blockquote></details>

LGTM
___
*[`490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java` line 18 at r1](https://reviewable.io/reviews/harness/harness-core/52325#gh-1312284362:-NdkI64731xeo3fnuPug:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/ba7e82d2d5e1c11742dfd3fe02ef214afea4b867/490-ce-commons/src/main/java/io/harness/ccm/commons/constants/ViewFieldConstants.java#L18)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Nothing just aligning all important cluster field constants together 
</blockquote></details>

LGTM
___
*[`batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/AnomalyDetectionTimeSeries.java` line 27 at r3](https://reviewable.io/reviews/harness/harness-core/52325#gh-1314831677:-NdkI9uKC_TPwwOGVmkt:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/0babfcac2a03508f3194a90a25b0c71804d58d9e/batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/AnomalyDetectionTimeSeries.java#L27)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Removed
</blockquote></details>

LGTM
___
*[`batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/helpers/TimeSeriesUtils.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/52325#gh-1314821109:-NdkIFL1-00tjQjnvxYg:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/0babfcac2a03508f3194a90a25b0c71804d58d9e/batch-processing/service/src/main/java/io/harness/batch/processing/anomalydetection/helpers/TimeSeriesUtils.java#L25)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Done
</blockquote></details>

LGTM
___
*[`batch-processing/service/src/main/java/io/harness/batch/processing/tasklet/support/HarnessEntitiesService.java` line 28 at r3](https://reviewable.io/reviews/harness/harness-core/52325#gh-1314834933:-NdkIHN6E0ZyUtttiHVi:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/0babfcac2a03508f3194a90a25b0c71804d58d9e/batch-processing/service/src/main/java/io/harness/batch/processing/tasklet/support/HarnessEntitiesService.java#L28)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Yes, but i don't think it provides any added advantage or make a difference
</blockquote></details>

LGTM
___
*[`ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/52325#gh-1312306818:-NdkIIiQCxhcyiZQod0H:bb6ywmx) ([raw file](https://github.com/harness/harness-core/blob/ba7e82d2d5e1c11742dfd3fe02ef214afea4b867/ce-nextgen/service/src/main/java/io/harness/ccm/migration/CENGTimescaleMigrationDetails.java#L44)):*
<details><summary><i>Previously, shashwat-harness wrote…</i></summary><blockquote>

Merged
</blockquote></details>

LGTM

**Comment by shashwat-harness** [comment]
trigger uts

**Comment by shashwat-harness** [comment]
trigger ut2

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger ut2

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by shashwat-harness** [comment]
trigger uts

**Comment by shashwat-harness** [comment]
trigger ut7

**Comment by rdsoze** [comment]
@shashwat-harness Going forward, please add more meaningful commit messages.

**Pull request merged**